### PR TITLE
Fix IDs and domain scoping

### PR DESCRIPTION
**Pull request by dadgar**

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -1342,32 +1342,30 @@ func (c *Client) setupConsulSyncer() error {
...
-			for taskName, taskState := range taskStates {
-				if taskState.State == structs.TaskStateRunning {
-					if tr, ok := ar.tasks[taskName]; ok {
-						for _, service := range tr.task.Services {
-							services = append(services, service)
-						}
-					}
-				}
+			if ar.Alloc().TerminalStatus() {
+				// Ignore non-running allocations
```
Comment (sean-) [inline client/client.go]: "Ignore non-terminal allocations"

Comment (sean-) [inline client/client.go]: Use old code here to iterate over tasks.

File: client/driver/executor/executor.go [hunk 2]
```diff
@@ -365,11 +362,25 @@ func (e *UniversalExecutor) UpdateTask(task *structs.Task) error {
 	// Re-syncing task with Consul agent
 	if e.consulSyncer != nil {
 		e.interpolateServices(e.ctx.Task)
-		e.consulSyncer.SetServices(e.ctx.AllocID, task.Services, executorServiceRegPrefix)
+		domain := consul.NewExecutorDomain(e.ctx.AllocID, task.Name)
+		serviceMap := servicesToServiceMap(e.ctx.AllocID, task.Services)
```
Comment (sean-) [inline client/driver/executor/executor.go]: `servicesToServiceMap()` -> `generateServiceKeys()`

File: client/driver/executor/executor.go [hunk 3]
```diff
@@ -365,11 +362,25 @@ func (e *UniversalExecutor) UpdateTask(task *structs.Task) error {
...
+		e.consulSyncer.SetServices(domain, serviceMap)
 	}
 	return nil
 }
 
+// servicesToServiceMap takes a list of interpolated services and returns a map
+// of ServiceKeys to services where the service key is appropriate for the
+// executor.
```
Comment (sean-) [inline client/driver/executor/executor.go]: Improve.  Nomad or Consul services?

File: command/agent/agent.go [hunk 4]
```diff
@@ -412,19 +414,20 @@ func (a *Agent) setupClient() error {
 
 	// Create the Nomad Server services for Consul
```
Comment (sean-) [inline command/agent/agent.go]: That comment is wrong, "Create the Nomad Client services for Consul."

File: command/agent/consul/syncer.go [hunk 5]
```diff
@@ -28,8 +52,8 @@ const (
 	// initialSyncDelay is the delay before an initial sync.
 	initialSyncDelay = 5 * time.Second
 
-	// nomadServicePrefix is the prefix used when registering a service
-	// with consul
+	// nomadServicePrefix is the first prefix that scopes all Nomad registered
+	// services
 	nomadServicePrefix = "nomad-registered-service"
```
Comment (sean-) [inline command/agent/consul/syncer.go]: Shorten to just `_nomad`

File: command/agent/consul/syncer.go [hunk 6]
```diff
@@ -910,28 +953,40 @@ func (c *Syncer) runCheck(check Check) {
...
-	servicesToKeep := make(map[string]struct{})
-	for allocID, services := range servicesInRunningAllocs {
-		for _, service := range services {
-			serviceID := c.GenerateServiceID(allocID, service, serviceRegPrefix)
-			servicesToKeep[serviceID] = struct{}{}
-		}
-	}
-
+// KeepDomains prunes all services that do not exist in the passed domains
+func (c *Syncer) KeepDomains(domains []ServiceDomain) error {
```
Comment (sean-) [inline command/agent/consul/syncer.go]: `Keep` -> `ReapUnmatched`

**Comment by sean-** [comment]
LGTM

File: client/client.go [hunk 7]
```diff
@@ -1342,32 +1342,30 @@ func (c *Client) setupConsulSyncer() error {
...
 	c.consulSyncer.AddPeriodicHandler("Nomad Client Fallback Server Handler", bootstrapFn)
 
-	// TODO this should only deregister things that the executors do not know
-	// about
-	consulServicesSyncFn := func() error {
-		const estInitialConsulServices = 8
-		const serviceGroupName = "executor"
-		servicesInRunningAllocs := make(map[string][]*structs.Service)
+	consulServicesReaperFn := func() error {
+		const estInitialExecutorDomains = 8
```
Comment (diptanu) [inline client/client.go]: Take this out of the function and add a comment?

**Pull request merged**